src/codegen/mod.rs

```diff
@@ -791,44 +788,127 @@ impl<'a> Bitfield<'a> {
-            last_field_align = cmp::max(last_field_align, width as usize);
+            last_field_align = cmp::max(last_field_align, width);
 
             unfilled_bits_in_last_unit = data_size - field_size_in_bits;
         }
 
         if field_size_in_bits != 0 {
-            // Push the last field.
-            let ty =
-                BlobTyBuilder::new(Layout::new(bytes_from_bits_pow2(field_size_in_bits),
-                                               bytes_from_bits_pow2(last_field_align)))
-                    .build();
-
-            let field = StructFieldBuilder::named(&last_field_name)
-                .pub_()
-                .build_ty(ty);
-            fields.push(field);
+            // Flush the last physical field and its bitfields.
+            fields.push(flush_bitfields(ctx,
+                                        field_size_in_bits,
+                                        last_field_align,
+                                        &last_field_name,
+                                        bitfields.drain(..),
+                                        methods));
         }
 
         Layout::new(bytes_from_bits(total_size_in_bits), max_align)
     }
 }
 
+/// A physical field (which is a word or byte or ...) has many logical bitfields
+/// contained within it, but not all bitfields are in the same physical field of
+/// a struct. This function creates a single physical field and flushes all the
+/// accessors for the logical `bitfields` within that physical field to the
+/// outgoing `methods`.
+fn flush_bitfields<'a, I>(ctx: &BindgenContext,
+                          field_size_in_bits: usize,
+                          field_align: usize,
+                          field_name: &str,
+                          bitfields: I,
+                          methods: &mut Vec<ast::ImplItem>) -> ast::StructField
+    where I: IntoIterator<Item = (&'a str, usize, usize, ast::Ty, Layout)>
+{
+    use aster::struct_field::StructFieldBuilder;
+
+    let field_layout = Layout::new(bytes_from_bits_pow2(field_size_in_bits),
+                                   bytes_from_bits_pow2(field_align));
+    let field_ty = BlobTyBuilder::new(field_layout).build();
+
+    let field = StructFieldBuilder::named(field_name)
+        .pub_()
+        .build_ty(field_ty.clone());
+
+    for (name, offset, width, bitfield_ty, bitfield_layout) in bitfields {
+        let prefix = ctx.trait_prefix();
+        let getter_name = ctx.rust_ident(name);
+        let setter_name = ctx.ext_cx()
+            .ident_of(&format!("set_{}", &name));
+        let field_ident = ctx.ext_cx().ident_of(field_name);
+
+        let field_int_ty = match field_layout.size {
+            8 => quote_ty!(ctx.ext_cx(), u64),
+            4 => quote_ty!(ctx.ext_cx(), u32),
+            2 => quote_ty!(ctx.ext_cx(), u16),
+            1 => quote_ty!(ctx.ext_cx(), u8),
+            _ => panic!("physical field containing bitfields should be sized \
```

Comment:
well, I guess someone could make bindgen panic with something like: `__u128 foo: 128` or something? If so, maybe we just want to skip the getter/setter?

(Not a big deal landing as-is for now)